### PR TITLE
DataRepository implementation

### PR DESCRIPTION
**Pull request by MagoDelBlocco**
This PR brings the DataRepository (#23) implementation to the project.

The data model used for this database is:
* primary key: commit hash
* kind: "revision" - for the entries relevant to this stage of the project
* fields:
*- timestamp
*- branch
*- builders (see #64)

This implementation relies heavily upon the API provided by the [Spring Cloud GCP](https://googleapis.dev/java/spring-cloud-gcp/1.2.2.RELEASE/index.html) framework, as such, the methods are fairly short. All inputs from the exposed methods are sanitized, to avoid undefined behaviour.

Adding/Updating entries is done with the help of the @ Entity annotated class BuildInfo (#63), as described in this [Spring Cloud GCP tutorial](https://spring.io/projects/spring-cloud-gcp#overview). All incoming classes are converted to BuildInfo, or a subclass of it, so that they can be directly added to the datastore, without any further conversions.

The getLastEntries method sorts all "revision" kind entries by timestamp, limiting the first "number" entries, starting from a given offset.

### CODE REVIEW

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ * replicate such a list without querying the database for it first, and the complexity
+ * of the indexing and querying would be great, without any apparent value.
+ *
+ * Useful links:
+ * - https://googleapis.dev/java/google-cloud-datastore/latest/index.html
+ * - https://googleapis.dev/java/spring-cloud-gcp/1.2.2.RELEASE/index.html
+ * - https://cloud.google.com/datastore/docs/concepts/
+ * - https://spring.io/projects/spring-cloud-gcp#overview
+ */
+@SpringBootApplication
```
Comment (holounic) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:51]: ```@SpringBootApplication``` marks the entry point of the SpringBoot app. Please consider changing it. Spring needs classes to be annotated to scan and execute them (besides POJOs, because the are just "data types" and don't execute anything).  [This article](https://www.baeldung.com/spring-component-repository-service#:~:text=Spring%20Annotations,-In%20most%20typical&text=Then%2C%20it%20registers%20each%20bean%20in%20the%20ApplicationContext.&text=%40Component%20is%20a%20generic%20stereotype,act%20as%20a%20database%20repository)  could be really helpful to choose which annotation to use in this case.

**Review by holounic — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java (hunk 1 above)
Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: Thank you for the heads up! I changed it to `@Repository` in [770dbe2](https://github.com/googleinterns/step240-2020/pull/66/commits/770dbe2eec3fe72cabbc543bc47e02ad760400c1). Let me know if there is anything else to solve regarding this annotation.

Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: Thank you!

**Review by holounic — APPROVED** [review]
LGTM!

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ * of the indexing and querying would be great, without any apparent value.
+ *
+ * Useful links:
+ * - https://googleapis.dev/java/google-cloud-datastore/latest/index.html
+ * - https://googleapis.dev/java/spring-cloud-gcp/1.2.2.RELEASE/index.html
+ * - https://cloud.google.com/datastore/docs/concepts/
+ * - https://spring.io/projects/spring-cloud-gcp#overview
+ */
+@Repository
+public class GCDataRepository implements DataRepository {
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:52]:
```suggestion
public class DatastoreRepository implements DataRepository {
```
The "GC" abbreviation in GCDataRepository is a bit unclear (naively I read it as "garbage-collected repository at first). I'd suggest using the name of the product (Google Datastore, or just Datastore) as a basis for the name of this class.

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.Date;
+import java.util.List;
+import org.springframework.boot.SpringApplication;
+import org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate;
+import org.springframework.stereotype.Repository;
+
+/**
+ * Implementation of the DataRepository, with the purpose of providing I/O to the Google Cloud
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:35]:
```suggestion
 * A DataRepository implementation backed up by Google Datastore.
```
(possibly followed by a link to Google Datastore docs)

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ * Datastore. Each individual database entry is modeled by the {@link #BuildInfo BuildInfo} class.
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:36]:
```suggestion
 * Datastore. Each database entry is modeled by the {@link #BuildInfo BuildInfo} class.
```
("each" together with "individual") is a bit redundant.

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+ * The relevant fields for the database, which are described in that class are:
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:37]:
```suggestion
 * The relevant fields for the database are:
```
(it's implied that this documentation is about the class)

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+ *    - Kind: "revision"
+ *    - Key: commit hash
+ * The database indexes after all the BuildInfo fields, besides the {@code builders} list.
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:40]:
```suggestion
 * The database indexes all BuildInfo fields, except the {@code builders} list.
```
([besides vs. except](https://english.stackexchange.com/questions/7984/difference-between-except-and-besides#:~:text=They%20have%20totally%20different%20meanings,besides%20serves%20that%20of%20inclusion.&text=The%20whole%20class%20is%20going%2C%20except%20John.&text=Besides%20the%20class%2C%20Mary%20is%20also%20going.))

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+ * It is unfeasable to query after the list of builders, since it would not be possible to
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:41]: I'd omit the reason why we don't index the builders list.
Effectively, we just not need to.
This brings up an additional point: do we really need indexes over all of the remaining fields of BuildInfo? I'd suggest starting only with the indexes required to implement the current DataRepository interface; we can add another index when we need it later.

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+ *
+ * Useful links:
+ * - https://googleapis.dev/java/google-cloud-datastore/latest/index.html
+ * - https://googleapis.dev/java/spring-cloud-gcp/1.2.2.RELEASE/index.html
+ * - https://cloud.google.com/datastore/docs/concepts/
+ * - https://spring.io/projects/spring-cloud-gcp#overview
+ */
+@Repository
+public class GCDataRepository implements DataRepository {
+  private DatastoreTemplate storage;
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:53]: Shouldn't we have also have a constructor that initializes this?
(c.f. Item 5 from "Effective Java")

**Review by krasimir-google — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java (hunk 2 above)
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:52]: How do we test this?

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java (hunk 8 above)
Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java:57]: Added `@Autowired` annotation to it. Hopefully, that would auto-initialize it, given that it has a `@Component` annotation. We'll have to wait for more PRs to merge to test that at runtime. Constructing it manually proved to be very complicated, since it required the Spring context and some other data of the `@Entity` classes.

Comment (krasimir-google) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: Thank you! I think `@Autowired` is fine.

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java (hunk 2 above)
Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: With #70, which will get merged in this branch

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java (hunk 7 above)
Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: This is a valid point. Thank you for bringing this up! I will also update #63 accordingly.

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java (hunk 6 above)
Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: removed the indexing part, according to the next comment

File: src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java (hunk 2 above)
Comment (krasimir-google) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: Great! I'd suggest to merge #70 into this one right away and keep working on the test cases here. Specifically, there is no need for a separate PR just for test cases. Keeping the implementation together with its tests together in a PR is one of the few exceptions to the "keep PRs small" rule.

Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/GCDataRepository.java]: In this case I will discard #70 directly, and update this PR further with the tests. Thank you for the clarification!

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  /**
+   * Tests the behaviour of add/update/get/delete requests on null data.
+   */
+  @Test
+  public void testRequestsNullData() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertFalse(storage.createRevisionEntry(null));
```
Comment (krasimir-google) [inline src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java:84]: I'd expect this to throw an `IllegalArgumentException` in this case, not return.
You can use `assertThrows` to set expectations around throwing exceptions: https://stackoverflow.com/questions/40268446/junit-5-how-to-assert-an-exception-is-thrown

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertTrue(storage.deleteRevisionEntry("2"));
+  }
+
+  /**
+   * Tests multiple threads creating update requests for the same entities.
+   */
+  @Test
+  public void testParallelBurstRequests() {
```
Comment (krasimir-google) [inline src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java:123]: I would not include a multithreaded unit test.
Unit tests like this are not the best way to test or set concurrency expectations. We have no control over how the Threads are going to be scheduled and how are they to interfere with each other, so effectively every run of the test could execute a very different execution. Often such tests become "flaky" (different executions pass or fail in the same configuration) which reduces their utility as a good way to prevent regressions.

File: src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+   */
+  public boolean createRevisionEntry(ParsedGitData entryData) throws IllegalArgumentException {
+    if (entryData == null) {
+      throw new IllegalArgumentException("entryData cannot be null");
+    }
+
+    if (getRevisionEntry(entryData.getCommitHash()) == null) {
+      try {
+        storage.save(new BuildInfo(entryData));
+      } catch (Exception e) {
```
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java:61]: This catch is very broad; I'd not catch it here but let it bubble up to the caller for now.
If an exception happens here, it's probably because of some connectivity or concurrence issue, so just returning `false` in that case potentially discards relevant exception information that may help with debugging. Alternatively, I'd suggest to log the exception so that we have a change to see and debug what happened.

Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java:61]: the .save(T instance) method uses the DatastoreReaderWriter::put method, which by the Java Reference states:
`put
Entity put(FullEntity<?> entity)
A Datastore put (a.k.a upsert) operation: inserts an entity if it does not exist, updates it otherwise. This method will automatically allocate an id if necessary.
Parameters:
entity - the entity to put
Returns:
an Entity with the same properties and a key that is either newly allocated or the same one if key is already complete
Throws:
DatastoreException - upon failure
IllegalArgumentException - if the given entity is missing a key`

So, basically in this case we can only theoretically get a DatastoreException or an IllegalArgumentException. Since we sanitize the inputs at every stage, the IllegalArgumentException would have been thrown earlier, which leaves only the DatastoreException. If I change `Exception` to `DatastoreException`, would it make sense to leave the rest of the code as-is?

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java (hunk 9 above)
Comment (MagoDelBlocco) [reply src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java:84]: This was an oversight on my part. It should indeed throw an exception, I'll patch it with the next commit. Thank you!

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java (hunk 10 above)
Comment (MagoDelBlocco) [reply src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java:123]: How can we test concurrent use of the datastore then?

Comment (krasimir-google) [reply src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java]: >How can we test concurrent use of the datastore then?

We don't.
We rely on the way datastore handles concurrent use itself. Concurrency guarantees are properties of the code that are not well exercised by traditional testing. Typically people would review the code and check if it acts within the specifications. In a more formal setting, it's possible to employ some formal proof techniques to have a mechanically verified proof of correctness (stuff like [TLA+](https://en.wikipedia.org/wiki/TLA%2B)), but that's not practical to do at the application code level as it's a lot of work. So usually those methods are employed for the code that actually implements the database layer, and the application layer just assumes well-behaved concurrency behavior from the database. In particular, Datastore enforces [serializable isolation](https://cloud.google.com/datastore/docs/concepts/transactions#isolation_and_consistency) and the code here should be well-behaved in such context.

File: src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java (hunk 11 above)
Comment (krasimir-google) [reply src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java:61]: Great!
> If I change Exception to DatastoreException, would it make sense to leave the rest of the code as-is?

That would be OK. I'd suggest still logging the DatastoreException in the catch clause in that case so we have a chance to see the details about such problems at runtime.

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java (hunk 10 above)
Comment (MagoDelBlocco) [reply src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java]: > Datastore enforces serializable isolation and the code here should be well-behaved in such context.

I believe this applies to Datastore transactions, which are a separate part of the Datastore API, since there is no mention of race conditions for regular operations, like add, delete or run. I have deleted this test for now, but I think it's worth testing if everything runs as expected in a parallel environment.

File: src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java (hunk 11 above)
Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java]: For now, the stack trace and exception + exception message are printed to stderr.

**Review by krasimir-google — APPROVED** [review]
LGTM
